src/picongpu/include/traits/UsesRNG.hpp

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2016 Marco Garten, Rene Widera
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <boost/mpl/bool.hpp>
+
+namespace picongpu
+{
+namespace traits
+{
+
+/** Checks if an object requires the RNG
+ *
+ * @tparam T_Object any object (class or typename)
+ *
+ * This struct must define
+ * ::type (boost::mpl::bool_<>)
+ */
+template<typename T_Object>
+struct UsesRNG
```

Comment:
I stand corrected: I forgot the `...::type` in the above example which turned it into unevaluated expressions not triggering the error.
As per offline discussion with @ax3l the solution is to inherit from `bmpl::bool_<true/false>` like in the following, corrected example:

    #include <boost/mpl/vector.hpp>
    #include <boost/mpl/copy_if.hpp>
    #include <boost/mpl/if.hpp>
    #include <boost/type_traits.hpp>

    namespace bmpl = boost::mpl;

    template<class T>
    struct Foo: public bmpl::bool_<true>{};

    template<>
    struct Foo<int>: public bmpl::bool_<false>{};

    typedef bmpl::vector<int, unsigned> MyVec;

    typedef typename bmpl::copy_if<MyVec, boost::is_unsigned<bmpl::_1> >::type Result;
    typedef typename bmpl::copy_if<MyVec, Foo<bmpl::_1> >::type Result2;
    typedef typename bmpl::copy_if<MyVec, bmpl::and_<Foo<bmpl::_1>, boost::is_unsigned<bmpl::_1> > >::type Result3;
    typedef typename bmpl::if_<bmpl::and_<Foo<int>, Foo<unsigned> >, int, bool>::type Result4;

    int main(){
        if(Foo<int>::value) return 1;
        else return 0;
    }